drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_tmc5041
+
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/drivers/stepper/stepper_trinamic.h>
+
+#include <zephyr/logging/log.h>
+
+#include "tmc_reg.h"
+#include "tmc_spi.h"
+
+LOG_MODULE_REGISTER(tmc5041, CONFIG_STEPPER_LOG_LEVEL);
+
+struct tmc5041_data {
+	struct k_mutex mutex;
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+	/** INT pin GPIO callback. */
+	struct gpio_callback int_cb;
+	/* Work item to run the callback in a thread context. */
+	struct k_work int_pin_callback_work;
+	/* device pointer required to access config in k_work */
+	const struct device *controller;
+#endif
+};
+
+struct tmc5041_config {
+	const uint32_t gconf;
+	struct spi_dt_spec spi;
+	const uint32_t clock_frequency;
+#if CONFIG_STEPPER_ADI_TMC5041_INT
+	struct gpio_dt_spec int_pin_spec;
+	/* connected motors */
+	const struct device **motors;
+	/* number of connected motors */
+	const uint8_t motor_count;
+#endif
+};
+
+struct tmc5041_stepper_data {
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	struct k_work_delayable stallguard_dwork;
+	/* device pointer required to access config in k_work */
+	const struct device *stepper;
+#endif
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+	struct k_poll_signal *async_signal;
+#endif
+};
+
+struct tmc5041_stepper_config {
+	const uint8_t index;
+	const uint16_t default_micro_step_res;
+	/* parent controller required for bus communication */
+	const struct device *controller;
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	int8_t stallguard_threshold;
+#endif
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	const struct tmc_ramp_generator_data default_ramp_config;
+#endif
+};
+
+static int tmc5041_write(const struct device *dev, const uint8_t reg_addr, const uint32_t reg_val)
+{
+	const struct tmc5041_config *config = dev->config;
+	struct tmc5041_data *data = dev->data;
+	const struct spi_dt_spec bus = config->spi;
+	int status = k_mutex_lock(&data->mutex, K_MSEC(100));
```

Comment:
Do you really want the timeout? you are ignoring the return value on all callers that I can tell and there's no warning here so this has no chance of timing out and being discovered

How about using `K_FOREVER` here, adding a `LOG_WRN` to the error handling and change the return to void and ditch all those `(void)` casting? Or change the callers to get the error code and log something.

Would clean the code below as well, normally you want to branch off error handling and keep the normal execution inline.